### PR TITLE
xrun: Introduce hardware configuration that can be accepted by xrun with unit tests

### PR DESCRIPTION
**Pull request by oleksiimoisieiev**
This PR was set to WIP until cmlient will support the new OCI spec.

xrun accepts json configuration, built using OCI spec to provide all
configuration to start the Domain.
Current implementation provides possibility to pass the real devices
to the Domains.
It can be configured using the following set of fields:
device-tree - domain device-tree that describes the devices that
should be used by the Domain;
dtdevs - list of the devices, that should be exposed to the Domains
partial device-tree;
vcpus - count of vcpus for the Domain;
mem_kb - Domain memory size if KB;
iomems - list of the iomems that should be exposed for the Domain;
irq - list of irqs that should be virtualized and used by the Domain.

### CODE REVIEW

File: tests/xrun/src/main.c [hunk 1]
```diff
@@ -0,0 +1,513 @@
...
+	int ret;
+	int i;
+	char buf[25];
+
+	test_json_contents = json;
+	test_dtb_contents = "dtb";
+	test_dtb_name = "uni.dtb";
+	test_image_name = "unikernel.bin";
+
+	for (i = 0; i < 200; i++) {
```
Comment (firscity) [inline tests/xrun/src/main.c:261]: I'm don't completely understand what's going on here, but I think the number of test runs should be configurable.

File: tests/xrun/src/mock-xen-dom-mgmt.c [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+
+#include <zephyr/tc_util.h>
+#include <zephyr/ztest.h>
+
+#include <xen_dom_mgmt.h>
+extern struct xen_domain_cfg g_cfg;
+
+int domain_create(struct xen_domain_cfg *domcfg, uint32_t domid)
+{
+	printk("%s: domid = %d\n", __func__, domid);
```
Comment (firscity) [inline tests/xrun/src/mock-xen-dom-mgmt.c:18]: Please, no `printk` except if it's required for test suite.

File: xrun/src/xrun.c [hunk 3]
```diff
@@ -477,7 +561,9 @@ int xrun_run(const char *bundle, int console_socket, const char *container_id)
 		}
 	}
 
-	ret = generate_cmdline(&spec, &container->cmdline);
+	memset(&domcfg, 0, sizeof(domcfg));
```
[the remark this answers is not in the document]
Comment (oleksiimoisieiev) [reply xrun/src/xrun.c]: fixed

File: xrun/src/xrun.c [hunk 4]
```diff
@@ -61,21 +78,23 @@ struct domain_spec {
 	struct vm_spec vm;
 };
 
+K_MUTEX_DEFINE(container_run_lock);
```
[the remark this answers is not in the document]
Comment (oleksiimoisieiev) [reply xrun/src/xrun.c]: +

File: tests/xrun/src/mock-xen-dom-mgmt.c (hunk 2 above)
Comment (oleksiimoisieiev) [reply tests/xrun/src/mock-xen-dom-mgmt.c]: fixed

File: tests/xrun/src/main.c (hunk 1 above)
Comment (oleksiimoisieiev) [reply tests/xrun/src/main.c]: The idea was to test massive container start/end routine. There was no intention to test xrun container limits. It shouls be tested in therms of xen-dom-mgmt. I've rewritten whole test to start container creation\removing in a couple of threads.

**Comment by firscity** [comment]
`Acked-by: Dmytro Firsov <dmytro_firsov@epam.com>`

File: xrun/src/xrun.c [hunk 5]
```diff
@@ -46,36 +49,52 @@ struct kernel_spec {
 	size_t params_len;
 };
 
+struct iomem_spec {
+	const long firstGFN;
```
Comment (al1img) [inline xrun/src/xrun.c:53]: uint64_t for all iomem fields

File: xrun/src/xrun.c [hunk 6]
```diff
@@ -46,36 +49,52 @@ struct kernel_spec {
...
+	const long firstMFN;
+	const long nrMFNs;
+};
+
 struct hwconfig_spec {
-	const char *devicetree;
+	const char *deviceTree;
+	const int vcpus;
```
Comment (al1img) [inline xrun/src/xrun.c:60]: uint32_t

File: xrun/src/xrun.c [hunk 7]
```diff
@@ -46,36 +49,52 @@ struct kernel_spec {
...
+	const int memKB;
+	const char *dtdevs[CONFIG_XRUN_DTDEVS_MAX];
+	const struct iomem_spec iomems[CONFIG_XRUN_IOMEMS_MAX];
+	const int irqs[CONFIG_XRUN_IRQS_MAX];
```
Comment (al1img) [inline xrun/src/xrun.c:64]: uint32_t for irqs

File: xrun/src/xrun.c [hunk 8]
```diff
@@ -25,10 +26,12 @@ LOG_MODULE_REGISTER(xrun);
 
 #define CONTAINER_NAME_SIZE 64
 #define UNIKERNEL_ID_START 12
+#define VCPUS_MAX_COUNT 24
 
 #define CONFIG_JSON_NAME "config.json"
 
 K_MUTEX_DEFINE(container_lock);
+
```
Comment (al1img) [inline xrun/src/xrun.c:34]: `static K_MUTEX_DEFINE(container_lock);`?

File: xrun/src/xrun.c [hunk 9]
```diff
@@ -486,24 +570,44 @@ int xrun_run(const char *bundle, int console_socket, const char *container_id)
...
 	LOG_DBG("domid = %lld", container->domid);
 
-	ret = fill_domcfg(container);
+	ret = fill_domcfg(&domcfg, &spec, container);
 	if (ret) {
 		goto err;
 	}
 
-	ret = domain_create(&container->domcfg, container->domid);
+	ret = domain_create(&domcfg, container->domid);
```
Comment (al1img) [inline xrun/src/xrun.c:593]: There is some discrepancy with mutexes usage. Does `domain_...` API's are thread-safe?

File: xrun/src/xrun.c (hunk 5 above)
Comment (oleksiimoisieiev) [reply xrun/src/xrun.c]: Fixed. Thakns.

File: xrun/src/xrun.c (hunk 6 above)
Comment (oleksiimoisieiev) [reply xrun/src/xrun.c]: +

File: xrun/src/xrun.c (hunk 7 above)
Comment (oleksiimoisieiev) [reply xrun/src/xrun.c]: +

File: xrun/src/xrun.c (hunk 9 above)
Comment (oleksiimoisieiev) [reply xrun/src/xrun.c:593]: We've discussed that offline. The idea was to use common sections and avoid complications with heap. Refactored a bit to lock only parts when globals are used.

File: xrun/src/xrun.c (hunk 8 above)
Comment (oleksiimoisieiev) [reply xrun/src/xrun.c:34]: fixed

**Comment by al1img** [comment]
Reviewed-by: Oleksandr Grytsov `<oleksandr_grytsov@epam.com>`

**Comment by oleksiimoisieiev** [comment]
@firscity GEt RB from @al1img . So please merge

**Comment by firscity** [comment]
@oleksiimoisieiev please add both my and @al1img tags to commits first.

**Pull request merged**